Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp

```diff
@@ -0,0 +1,239 @@
+            return;
+        }
+    }
+
+    RELEASE_ASSERT(m_resourceBaseURL.protocolIsFile());
+    RELEASE_ASSERT(m_resourceBaseURL.hasPath());
+
+    if (m_resourceBaseURL.path().right(1) != "/"_s)
+        m_resourceBaseURL = URL::fileURLWithFileSystemPath(FileSystem::pathByAppendingComponent(m_resourceBaseURL.path(), "/"_s));
+
+    if (!manifestParsedSuccessfully()) {
+        ASSERT(!m_errors.isEmpty());
+        outError = m_errors.last().ptr();
+    }
+}
+
+WebExtension::WebExtension(const JSON::Value& manifest, Resources&& resources)
+    : m_manifestJSON(manifest)
+    , m_resources(WTFMove(resources))
+{
+    auto manifestString = manifest.toJSONString();
+    RELEASE_ASSERT(manifestString);
+
+    m_resources.set("manifest.json"_s, manifestString);
+}
+
+Expected<Ref<API::Data>, RefPtr<API::Error>> WebExtension::resourceDataForPath(const String& originalPath, CacheResult cacheResult, SuppressNotFoundErrors suppressErrors)
+{
+    ASSERT(originalPath);
+
+    String path = originalPath;
+
+    // Remove leading slash to normalize the path for lookup/storage in the cache dictionary.
+    if (path.startsWith('/'))
+        path = path.substring(1);
+
+    if (path.startsWith("data:"_s)) {
+        if (auto base64Position = path.find(";base64,"_s); base64Position != notFound) {
+            auto base64String = base64DecodeToString(path.substring(base64Position));
+            return API::Data::create(base64String.utf8().span());
+        }
+
+        if (auto commaPosition = path.find(','); commaPosition != notFound) {
+            auto urlEncodedString = path.substring(commaPosition);
+            auto decodedString = URL(urlEncodedString).string();
+            return API::Data::create(decodedString.utf8().span());
+        }
+
+        ASSERT(path == "data:"_s);
+        return API::Data::create(std::span<const uint8_t> { });
+    }
+
+    if (path == generatedBackgroundPageFilename || path  == generatedBackgroundServiceWorkerFilename)
+        return API::Data::create(generatedBackgroundContent().utf8().span());
+
+    if (auto entry = m_resources.find(path); entry != m_resources.end()) {
+        return WTF::switchOn(entry->value,
+            [](const Ref<API::Data>& data) {
+                return data;
+            },
+            [](const String& string) {
+                return API::Data::create(string.utf8().span());
+            });
+    }
+
+    auto resourceURL = resourceFileURLForPath(path);
+    if (resourceURL.isEmpty()) {
+        if (suppressErrors == SuppressNotFoundErrors::No)
+            return makeUnexpected(createError(Error::ResourceNotFound, WEB_UI_FORMAT_STRING("Unable to find \"%s\" in the extension’s resources. It is an invalid path.", "WKWebExtensionErrorResourceNotFound description with invalid file path", path.utf8().data())));
+        return makeUnexpected(nullptr);
+    }
+
+    auto rawData = FileSystem::readEntireFile((resourceURL).fileSystemPath());
```

Comment:
Unneeded parenthesis:

```suggestion
    auto rawData = FileSystem::readEntireFile(resourceURL.fileSystemPath());
```

Or am I missing something?